0.0.1/agent.py

```diff
@@ -1,93 +1,188 @@
+* Tell the user if you don't support a capability. Do NOT make up or provide false information or figures.
+* Do not expose the functions you have access to to the user.
+* Do not use figures or function call from preceding messages to generate responses.
+""",
+        }
 
         # Use the model set in the metadata to generate a response
-        result = self.env.completion([prompt] + self.env.list_messages())
+        result = None
+        tools = self.env.get_tool_registry().get_all_tool_definitions()
+        tools_completion = self.env.completion_and_get_tools_calls(
+            [prompt] + [self.env.get_last_message() or ""], tools=tools
+        )
+        self.env.add_system_log(
+            f"Should call tools: {tools_completion.tool_calls}", logging.DEBUG
+        )
+
+        if tools_completion.message:
+            self.env.add_reply(tools_completion.message)
+        if tools_completion.tool_calls and len(tools_completion.tool_calls) > 0:
+            tool_call_results = self._handle_tool_calls(tools_completion.tool_calls)
+            if len(tool_call_results) > 0:
+                self.env.add_system_log(
+                    f"Got tool call results: {tool_call_results}", logging.DEBUG
+                )
+
+                context = [prompt] + self.env.list_messages() + tool_call_results
+                result = self.env.completion(context)
 
-        chat_history = self.env.list_messages()
-        last_user_query = self.get_last_search_term(chat_history)
-        if last_user_query == "recommend swaps" or last_user_query == "suggest tokens to swap":
-          result = self.recommend_token_allocations_to_swap_for_stablecoins()
-        elif last_user_query == "show allowance goal":
-          result = self.get_allowance_goal()
-        elif last_user_query == "show growth goal":
-          result = self.get_growth_goal()
-        elif last_user_query == "fetch prices":
-          result = self.fetch_token_prices()
+                self.env.add_system_log(
+                    f"Got completion for tool call with results: {result}. Context: {context}",
+                    logging.DEBUG,
+                )
 
-        self.env.add_reply(result)
+        if result:
+            self.env.add_reply(result)
 
         # Give the prompt back to the user
         self.env.request_user_input()
 
-    def find_near_account_id(self):
-        """Save the NEAR account ID of the user from chat history format 'near: <account_id>'"""
+    @staticmethod
+    def _to_function_response(function_name: str, value: typing.Any) -> typing.Dict:
+        """
+        Use to tell the LLM the result from a function call in a structured way
+        """
+        return {
+            "role": "function",
+            "name": function_name,
+            "content": json.dumps(value),
+        }
+
+    def _get_tool_name(self) -> str:
+        """Return the function name of the calling function as the tool name"""
+        return inspect.stack()[1][3]
```

Comment:
@RyanSept Am I correct to assume this is for our benefit that we could log/record as a way to understand which prompts are triggering which functions?

Reply:
This is mostly used when constructing responses for function tool calls. We need it to get the tool name when constructing a structured response after it's called via the `_to_function_response(func_name, return_value)` function in this class